File: Machine_Learning_A_Z/Part 1 - Data Preprocessing/1 - DataPreprocessing.py
#DATA PREPROCESSING

#IMPORTING LIBRARIES
import numpy as np 
import matplotlib.pyplot as plt
import pandas as pd

#IMPORTING THE DATASET
dataset = pd.read_csv('Data.csv')

#CREATING HTE MATRIX OF FEATURES
x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 3].values
x = pd.DataFrame(x)
y = pd.DataFrame(y)

x.columns = ['Country', 'Age', 'Salary']
y.columns = ['Purchased']



#------------------------------------DEALING WITH MISSING DATA------------------------------------#
from sklearn.preprocessing import Imputer

imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)

imputer = imputer.fit(x.values[:, 1:3])
x.values[:, 1:3] = imputer.transform(x.values[:, 1:3])



#------------------------------------ENCODING CATEGORICAL VARIABLES------------------------------------#
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

#FOR INDEPENDENT VARIABLE
labelencoder_x = LabelEncoder()
x.values[:, 0] = labelencoder_x.fit_transform(x.values[:, 0])
#CREATING DUMMY VARIABLES FOR CATEGORICAL VARIABLE
onehotencoder = OneHotEncoder(categorical_features = [0])
x = onehotencoder.fit_transform(x).toarray()

#DUMMY VARIABLE TRAP
#x = x[:, 1:]

#FOR DEPENDENT VARIABLE
labelencoder_y = LabelEncoder()
y.values[:, 0] = labelencoder_y.fit_transform(y.values[:, 0])




#------------------------------------SPLITTING INTO TRAIN AND TEST DATASET------------------------------------#
from sklearn.cross_validation import train_test_split

#from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)




#------------------------------------FEATURE SCALING------------------------------------#
# 1. STANDARDIZATION 
# 2. NORMALIZATION

from sklearn.preprocessing import StandardScaler
scl_x = StandardScaler()
x_train = scl_x.fit_transform(x_train)
x_test = scl_x.transform(x_test)

File: Machine_Learning_A_Z/Part 1 - Data Preprocessing/1 - DataPreprocessing.R
#DATA PREPROCESSING

getwd()
setwd("/Users/arajnarain/Desktop/Udemy/Machine Learning/Machine Learning A-Z Template Folder/Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/Data_Preprocessing")

dataset <- read.csv('Data.csv')

#------------------------------------DEALING WITH MISSING DATA------------------------------------#
dataset[is.na(dataset$Age),"Age"] <- mean(dataset[,"Age"], na.rm = T)
dataset[is.na(dataset$Salary), "Salary"] <- mean(dataset[,"Salary"], na.rm = T)
#OR
dataset$Age <- ifelse(is.na(dataset$Age),
                      ave(dataset$Age, 
                          FUN =  function(x) mean(x, na.rm = T)),
                      dataset$Age)
dataset$Salary <- ifelse(is.na(dataset$Salary),
                        ave(dataset$Salary, 
                            FUN =  function(x) mean(x, na.rm = T)),
                        dataset$Salary)



#------------------------------------ENCODING CATEGORICAL VARIABLES------------------------------------#
dataset$Country <- factor(dataset$Country,
                          levels = c("France", "Spain", "Germany"),
                          labels = c(1, 2, 3))
dataset$Purchased <- factor(dataset$Purchased,
                            levels = c("No", "Yes"),
                            labels = c(0, 1))
str(dataset)



#------------------------------------SPLITTING INTO TRAIN AND TEST DATASET------------------------------------#
#install.packages('caTools')
library(caTools)

set.seed(123)

split <- sample.split(dataset$Purchased, SplitRatio = 0.8)

training_set <- dataset[split, ]
test_set <- dataset[!split, ]
#OR
training_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)



#------------------------------------FEATURE SCALING------------------------------------#
training_set[, 2:3] <- scale(training_set[, 2:3])
test_set[, 2:3] <- scale(test_set[, 2:3])
